# Exam 1

Instructions:  

You are allowed to use the following materials and nothing else on this exam:  

1. the textbook and 
2. your flashcards 
    
on this exam.  


You are not allowed to use another person or any internet resources. This exam is not timed and you are free to start and stop as you please.  Good Luck, you'll do great.

Name: _________________
    

    
## Short Written Answer
1. What is the keyboard shortcut for turning a text cell into a code cell in a Jupyter notebook.

Answer: Y

2. What type of number is produced when an integer is multiplied by another integer?

Answer: Integer

3. What does the `//` operator do? (i.e. `5//2`)

Answer: Int division

4. How do you raise a number to a power in python?

Answer: **

5. What happens when you "add" two strings together?

Answer: They are placed next to each other

6. What happens when you multiply two strings together?

Answer: Error

7. How do you write a number in scientific notation in Python?

Answer:  numberenumber 
$3*10^8$ is 3e8

8. What does the `int` function do?

Answer: Converts a string or float to an integer

9. What does `a -= 10` mean?

Answer: a is now defined as what a was minus 10

10. What is the correct way to check to see if two **float variables** are equal to each other?  (You can write some code to illustrate if you prefer.)

Answer: abs(float1 - float2) < 1e-8 

# Short Code Answer
1. Extract the first through the 8th elements in the list below.

In [88]:
myList = [5,2,1,4,3,5,1,2,4,1,3,5,9,6,5]
myList[:8]

[5, 2, 1, 4, 3, 5, 1, 2]

2. Write a **single line of code** that prints `True` if $x$ is greater than $10$ **and** $y$ is less than $50$. Otherwise, it prints `False`.

In [89]:
x = 12
y = 32
x>10 and y < 50

True

3. Add the following electron configuration to the end of the list shown below: "1s2-2s2-2p6-3s1"

In [90]:
electrons = ["1s1","1s2", "1s2-2s1","1s2-2s2","1s2-2s2-2p1","1s2-2s2-2p2","1s2-2s2-2p3","1s2-2s2-2p4","1s2-2s2-2p5","1s2-2s2-2p6"]
electrons.append("1s2-2s2-2p6-3s1")

4. Using a single line of code, count the number of string that have "2s2" in them.

In [91]:
electrons = ["1s1","1s2", "1s2-2s1","1s2-2s2","1s2-2s2-2p1","1s2-2s2-2p2","1s2-2s2-2p3","1s2-2s2-2p4","1s2-2s2-2p5","1s2-2s2-2p6"]
electrons.count("2s2")

0

5. Find the maximum value in the list of numbers in the cell below.

In [92]:
from numpy.random import uniform
x = uniform(0,10,1000)
max(x)

9.997776984111566

6. Sum up all of the numbers in the list below.

In [93]:
from numpy.random import uniform
x = uniform(0,10,1000)
sum(x)

4991.093591600095

7. Using a single line of code, delete all of the repeated numbers in the list below.  How long is the resulting list?

In [94]:
from numpy.random import randint
x = randint(0,10,1000)
[(set(x))]

[{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}]

8. Create a dictionary that contains the ages of 5 of your family members.  Let the names be the keys and the ages the values.

In [95]:
# Python code here 
ages = {"mom":51,"dad":56,"Brenden":21,"Kallie":22,"Sadie":18}

9. The cell below contains a dictionary with the orbital radii of the planets in our solar system.  Use a single line of code to produce a list that contains all of the planet's names.

In [96]:
orbitalRadii = {"Mercury":0.39, "Venus":0.72, "Earth":1.0, "Mars":1.5,"Jupiter":5.2,"Saturn":9.5, "Uranus":19.2,"Neptune":30.1, "Pluto": 39.5}
list = orbitalRadii.keys()

10. The cell below contains a dictionary with the orbital radii of the planets in our solar system.   Use a single line of code to remove the entry for Mars.

In [97]:
orbitalRadii = {"Mercury":0.39, "Venus":0.72, "Earth":1.0, "Mars":1.5,"Jupiter":5.2,"Saturn":9.5, "Uranus":19.2,"Neptune":30.1, "Pluto": 39.5}
orbitalRadii.pop("Mars")

1.5

11. The cell below contains two sets.  Find their intersection.

In [98]:
squares= {1,4,9,16,25,36,49,64,81,100,121,144,169,196,225}
triangles = {1, 3, 6, 10, 15, 21, 28, 36, 45, 55, 66, 78, 91, 105, 120, 136, 153, 171, 190}
print(squares&triangles)

{1, 36}


12. The cell below contains a list of numbers.  Calculate the average of those numbers using a single line of code.

In [99]:
from numpy.random import uniform
from numpy import mean

x = uniform(0,10,1000)
print(mean(x))

avg = sum(x)/len(x)
print(avg)

4.963565428083231
4.963565428083243


## Code Answer

1. Calculate the following quantities in the cell provided below.
    1. $\cos({\pi \over 4})$
    2. $\sin(136^\circ)$
    3. $\ln(55)$

In [100]:
# Python code here.
from numpy import cos,pi,sin,log

a = cos(pi/4)
b = sin(136*pi/180)
c = log(55)

2. Perform the following operations in the cell below:
    1. Create a variable and assign its value to be 5.2.
    2. Use augmented assignment to multiply the variable by 14.
    3. Use augmented assignment to subtract 22 from the variable.
    4. Convert the variable into an integer
    5. Check to see if this variable is a multiple of 3
    6. Print the variable inside the following sentence: "This is the answer:" followed by the value of the variable.  Format the variable to display with only two numbers after the decimal.

In [101]:
# Python code here.
a = 5.2
a *= 14
a -= 22
a = int(a)
d = a % 3 == 0
print(d)
print(f'This is the answer. {a:.2f}')

False
This is the answer. 50.00


3. In the cell below, you'll find some strings containing the electronic configuration of the first 11 elements on the periodic table.  
    1. Using a single line of code, combine the strings to form one long string.
    2. Using a single line of code, determine the total length of the long string.
    3. Extract the substring starting at the 6th element going up to and including the 13th element.
    4. Now split the string on every occurrence of "-", producing a list of 11 separate strings.
    5. Extract the electron configuration for Boron (B) from the list that was produced in step 4. 


In [102]:
H = "1s1"
He = "-1s2"
Li = "-[He]2s1"
Be = "-[He]2s2"
B = "-[He]2s2_2p1"
C = "-[He]2s2_2p2"
N = "-[He]2s2_2p3"
O = "-[He]2s2_2p4"
F = "-[He]2s2_2p5"
Ne = "-[He]2s2_2p6"
Na = "-[Ne]3s1"
long_string = H+He+Li+Be+B+C+N+O+F+Ne+Na
len(long_string)
long_string[6:14]
long_list=long_string.split("-")
long_list.pop(long_list.index(B[1:]))
print(long_list)

['1s1', '1s2', '[He]2s1', '[He]2s2', '[He]2s2_2p2', '[He]2s2_2p3', '[He]2s2_2p4', '[He]2s2_2p5', '[He]2s2_2p6', '[Ne]3s1']


4. Write a short code that uses a given wind speed and accomplishes the following:

    1. Determines whether the storm qualifies as a hurricane and 
    2. If it does qualify as a hurricane, determine whether it is a category 1,2,3,4, or 5 storm.


| Category | Wind Speed (mph)|
|----------|-----------------|
|1         |  74-95          |
|2         |  76-110          |
|3         |  111-130          |
|4         |  131-155          |
|5         |  156-above          |

In [103]:
windSpeed = 80
if windSpeed >= 74:
    hurricane = True
    if windSpeed >= 156:
        category = 5
    elif windSpeed >=131:
        category = 4
    elif windSpeed >= 111:
        category = 3
    elif windSpeed >= 96:
        category = 2
    elif windSpeed >= 74:
        category = 1
    
    print(f'This is a category {category} hurricane')
else:
    hurricane = False
    print('This is not a hurricane')

This is a category 1 hurricane


5. Construct a dictionary containing the names of nobel prize in physics recipients from the years 1910 - 1925. (see [here](https://www.britannica.com/topic/Winners-of-the-Nobel-Prize-for-Physics-1856942) ) The year should be the key and the name(s) of the recipient(s) should be the value.

In [104]:

# Python code here.

6. Construct a list of the first 1000 numbers that are both a multiple of 7 and are even numbers.

In [105]:
numbers = []
num = 1
while len(numbers) < 1000:
    if num % 7 ==0 and num % 2 ==0:
        numbers.append(num)
    num +=1

7. Use a loop to calculate the following summation for $x = 0.5$: $$ \sum_{n = 0}^\infty x^n$$  Do it three different ways:
    1. Using a `for` loop with enough iterations that the sum isn't changing much. 
    2. Using a `while` loop that continues until $x^n < 1 \times 10^{-4}$.
    3. Using list comprehension and the `sum` function. (Use the same number of terms that you used for your `for` loop.)
    
The sum should converge to $${1 \over 1- x}$$ if you include a sufficient number of terms.  Check that it does.

In [106]:
# Python code here.
x = 0.5
summation = 0
for n in range(1000):
    summation += x**n
print(summation)

summation = 0
n = 0
while x**n > 1e-4:
    summation += x**n
    n += 1
print(summation) 

list_comp = [x**n for n in range(1000)]
print(sum(list_comp))

2.0
1.9998779296875
2.0


8. Kepler's third law relates a celestial body's orbital radius to its orbital period via the following equation $$T^2 = \left({4 \pi^2 \over G M_S} \right) r^3$$  where $M_S = 1.99 \times 10^{30}$ kg (mass of the sun)  and $r$ is the orbital radius of the planet

    1. Construct a function that takes the mass of the orbital radius ($r$) of the planet as an argument and returns its orbital period ($T$).  The gravitation constant ($G = 6.67 \times 10^{-11}$) should be included as a **keyword argument with a default value attached**.
    2. Now use list comprehension (or a traditional loop) to call your function for all of the planets in our solar system and construct a list containing their orbital periods. Convert these periods to Earth days. The orbital radii of the planets is given in the table below:
    
   | Planet | Orbital Radius (m)|
   |--------|---------------|
   | Mercury |$$5.79 \times 10^{10}$$|
   | Venus |$$1.08 \times 10^{11}$$|
   | Earth |$$1.50 \times 10^{11}$$|
   | Mars |$$2.28 \times 10^{11}$$|
   | Jupiter |$$7.79 \times 10^{11}$$|
   | Saturn |$$1.43 \times 10^{12}$$|
   | Uranus |$$2.87 \times 10^{12}$$|
   | Neptune| $$4.52 \times 10^{12}$$|
   | Pluto |$$5.91 \times 10^{12}$$|

In [107]:
# Python code here.
from numpy import pi,sqrt

def period(r,M_s = 1.99e30, G = 6.67e-11):
    return sqrt(4*pi**2/(G*M_s) * r**3)

radii = [5.79e10,1.08e11,1.5e11,2.28e11,7.79e11,1.43e12,2.87e12,4.52e12,5.91e12]
periods = [period(r)/60/60/24 for r in radii]
print(periods)

[87.94163107147473, 224.03315962690047, 366.70223699729064, 687.1933390185086, 4339.931324266559, 10793.958075794199, 30690.18566825252, 60657.47430716834, 90689.68470950938]


9. Perform the following:
    1. Construct a list of integers starting at $1$ and ending at $100$ (inclusive).
    2. Use list slicing to extract just the odd numbers from this list.
    3. Use list comprehension to square all of the remaining numbers.
    4. Use a list method to find the location of the number $361$ in the list.
    5. Add the number $101^2$ to the end of the list.
    6. Add the numbers $[103^2, 105^2]$ to the end of the list using a single line of code.
    7. Reverse the order of the list.
    8. Add up all of the values and print the result.


In [108]:
# Python code here.
start = range(1,101)
part2 = start[::2]
newList = [n**2 for n in part2 ]
newList.index(361)
newList.append(101**2)
newList.extend([103**2,105**2])
newList.reverse()
print(sum(newList))

198485


10. In the cell below you will find two lists containing the masses and velocities of several objects.  Use a loop to calculate the kinetic energy of each object using the following equation. $$K = {1\over 2} m v^2$$ Store the kinetic energies in a list. Then calculate the max, min, and average value of the energies.

In [109]:
from numpy.random import uniform

def kinetic_energy(m,v):
    return 1/2 * m * v**2

mass = uniform(5,10,1000)
velocity = uniform(8,23,1000)

kineticEnergy = []
for idx,m in enumerate(mass):
    K = kinetic_energy(m,velocity[idx])
    kineticEnergy.append(K)
    
maxm=max(kineticEnergy)
minm=min(kineticEnergy)
avg = sum(kineticEnergy)/len(kineticEnergy)
print(maxm,minm,avg)

2558.9661495350574 163.7575478507328 977.229961059983
